C++ IR: support for translating ErrorExpr

jbj opened the pull request:
Real-world snapshots contain `ErrorExpr` expressions where the extractor has detected an error in the code because it doesn't perfectly mimic the compiler. If these are not translated to something in the IR, it will lead to loose ends both in successors and operands.

jbj commented:
I've translated `ErrorExpr` to `NoOp` instructions, which is not ideal. All other `NoOp` instructions return `void`, but these can have other types. Do we have an existing opcode we can use, or do we need to invent something new?

rdmarsh2 commented:
We don't have an appropriate existing opcode, so we'll need to invent a new one.

rdmarsh2 commented:
Also, what semantics should `ErrorExpr` have? In particular, should we assume an `ErrorExpr` could potentially write to any virtual variable in the function?

jbj commented:
We could make an opcode that produces an arbitrary value of a given type when evaluated. Perhaps that opcode is also suitable for the unknown values that are stored to variables appearing in inline assembly statements. It could even be used for the initial value stored to an uninitialised variable, although we'd need to distinguish between arbitrary values that are due to limitations in the analysis vs. uninitialised data that can potentially cause bugs in the program. An arbitrary-value opcode also seems related to side effect instructions -- a call side effect is like storing an arbitrary value at the address of a function argument.

This was just a brainstorm. @dave-bartolomeo, what are your thoughts?

dave-bartolomeo commented:
@jbj I think there's an important distinction between `Uninitialized` (reading the result value provokes undefined behavior) and `CallSideEffect` (reading the result value produces a well-defined value, but we may not know much about the possible values due to analysis limitations). The result of an `ErrorExpr` probably falls into a third category. These three opcodes may all have a similar form, but they are semantically different.

jbj commented:
I've rebased to fix conflicts and added a commit to introduce `Opcode::Error` and use that. I remember now that you've convinced me it's fine to have many opcodes as long as we group similar opcodes using the `Instruction` class hierarchy. It's possible that the new `ErrorInstruction` could be grouped together with other instructions that produce well-defined but unknown values, but I've left that as future work since it's not blocking the IR deployment.

> Also, what semantics should `ErrorExpr` have? In particular, should we assume an `ErrorExpr` could potentially write to any virtual variable in the function?

It could also potentially write to every global value or field, including ones that aren't mentioned in the function. I think that modelling `ErrorExpr` soundly is too much work for too little gain. I've instead written some advice in the QLDoc for `ErrorInstruction`.

dave-bartolomeo commented on cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll:
```diff
@@ -2862,3 +2862,36 @@ class TranslatedStmtExpr extends TranslatedNonConstantExpr {
+    // The extractor deliberately emits an `ErrorExpr` as the first argument to
+    // the allocator call, if any, of a `NewOrNewArrayExpr`. That `ErrorExpr`
+    // should not be translated.
+    not exists(NewOrNewArrayExpr new | expr = new.getAllocatorCall().getArgument(0))
```
I believe we usually filter out skipped expressions in TranslatedElement.ql.
